kernel/src/init.rs

```diff
@@ -0,0 +1,182 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Kernel initialization.
+
+use component::InitStage;
+use ostd::{
+    arch::qemu::{exit_qemu, QemuExitCode},
+    boot::boot_info,
+    cpu::CpuId,
+};
+use spin::once::Once;
+
+use crate::{
+    fs::{fs_resolver::FsResolver, path::MountNamespace},
+    kcmdline::KCmdlineArg,
+    prelude::*,
+    process::{spawn_init_process, Process},
+    sched::SchedPolicy,
+    thread::kernel_thread::ThreadOptions,
+};
+
+// Kernel entry point.
+#[ostd::main]
+#[controlled]
+fn main() {
+    // Initializes the global states for all CPUs.
+    ostd::early_println!("[kernel] OSTD initialized. Preparing components.");
+    component::init_all(InitStage::Bootstrap, component::parse_metadata!()).unwrap();
+    init();
+
+    // Initializes the per-CPU states for BSP.
+    init_on_each_cpu();
+
+    // Enable APs.
+    ostd::boot::smp::register_ap_entry(ap_init);
+
+    // Give the control of the BSP to the idle thread.
+    ThreadOptions::new(bsp_idle_loop)
+        .cpu_affinity(CpuId::bsp().into())
+        .sched_policy(SchedPolicy::Idle)
+        .spawn();
+}
+
+fn init() {
+    crate::thread::init();
+    crate::util::random::init();
+    crate::driver::init();
+    crate::time::init();
+    crate::net::init();
+    crate::sched::init();
+    crate::process::init();
+    crate::fs::init();
+}
+
+fn init_on_each_cpu() {
+    crate::sched::init_on_each_cpu();
+    crate::process::init_on_each_cpu();
+    crate::fs::init_on_each_cpu();
+    crate::time::init_on_each_cpu();
+}
+
+fn ap_init() {
+    init_on_each_cpu();
+
+    ThreadOptions::new(ap_idle_loop)
+        // No races because `ap_init` runs on a certain AP.
+        .cpu_affinity(CpuId::current_racy().into())
+        .sched_policy(SchedPolicy::Idle)
+        .spawn();
+}
+
+//--------------------------------------------------------------------------
+// The first kernel thread
+//--------------------------------------------------------------------------
```

Comment:
I think the "the first kthread" section should be moved below the "the per-CPU idle threads" section. The entry points spawn the idle threads, which then spawn the first kthread. The current layout is slightly confusing because I have to go to the end of the file, then go back to the middle, to understand the logic.